app/src/heartbeat_info_desktop.cc

```diff
@@ -16,87 +16,22 @@
 
 #include "app/src/heartbeat_info_desktop.h"
 
-#include "app/src/heartbeat_date_storage_desktop.h"
-#include "app/src/log.h"
+#include <string>
 
-namespace firebase {
-namespace {
-
-const std::time_t kHeartbeatIntervalSeconds = 86400;  // 24 hours
-const char kHeartbeatGlobalTag[] = "GLOBAL";
-
-// Heartbeat may be invoked many times consecutively; make sure to only log an
-// error once to avoid spamming the log.
-void LogOnce(const std::string& message) {
-  static bool error_logged = false;
-  if (!error_logged) {
-    error_logged = true;
-    LogInfo(message.c_str());
-  }
-}
-
-// Returns whether the heartbeat for the SDK referred to by `tag` needs to be
-// sent to the backend.
-bool CheckAndUpdateHeartbeatTime(const char* tag,
-                                 HeartbeatDateStorage& storage) {
-  // Will return 0 if `tag` isn't in the map yet.
-  std::time_t last_heartbeat_seconds = storage.Get(tag);
-
-  std::time_t current_time_seconds = 0;
-  std::time(&current_time_seconds);
-  if (current_time_seconds - last_heartbeat_seconds <
-      kHeartbeatIntervalSeconds) {
-    return false;
-  }
-
-  storage.Set(tag, current_time_seconds);
-
-  return true;
-}
-
-bool ReadFromStorage(HeartbeatDateStorage& storage) {
-  bool ok = storage.ReadPersisted();
-  if (!ok) {
-    LogOnce("Heartbeat failed: unable to read the heartbeat data");
-  }
-  return ok;
-}
-
-bool WriteToStorage(const HeartbeatDateStorage& storage) {
-  bool ok = storage.WritePersisted();
-  if (!ok) {
-    LogOnce("Heartbeat failed: unable to write the heartbeat data");
-  }
-  return ok;
-}
-
-}  // namespace
+#include "app/memory/shared_ptr.h"
+#include "app/src/heartbeat/heartbeat_controller_desktop.h"
+#include "app/src/include/firebase/app.h"
 
-HeartbeatInfo::Code HeartbeatInfo::GetHeartbeatCode(const char* tag) {
-  HeartbeatDateStorage storage;
-
-  bool read_ok = ReadFromStorage(storage);
-  if (!read_ok) {
-    return Code::None;
-  }
-
-  bool send_sdk = CheckAndUpdateHeartbeatTime(tag, storage);
-  bool send_global = CheckAndUpdateHeartbeatTime(kHeartbeatGlobalTag, storage);
-
-  bool write_ok = WriteToStorage(storage);
-  if (!write_ok) {
-    return Code::None;
-  }
+namespace firebase {
 
-  if (!send_sdk && !send_global) {
-    return Code::None;
-  } else if (send_sdk && !send_global) {
-    return Code::Sdk;
-  } else if (!send_sdk && send_global) {
+HeartbeatInfo::Code HeartbeatInfo::GetHeartbeatCode(
+    SharedPtr<heartbeat::HeartbeatController> heartbeat_controller) {
+  std::string payload =
+      heartbeat_controller->GetAndResetTodaysStoredHeartbeats();
```

Comment:
I expect the content of the HeartbeatController's payload to be irrelevant. I was just using the payload to determine whether or not a heartbeat was logged.

As for which user agents are attached, I assumed that the existing logic would be unchanged, which is that firebase metadata provider logs the following:
  context.AddMetadata(kXFirebaseClientHeader, App::GetUserAgent());
  
I suppose there is a possibility to improve the firestore heartbeat logging, so instead of reading user agents at the time that the request is created, it could instead get user agents from the returned heartbeat payload. Is this what you were expecting?